### PR TITLE
Issue #1304 - Support parenthetical expressions in constraint language

### PR DESCRIPTION
**Pull request by MaxMcAdam**

### CODE REVIEW

File: externalpolicy/text_language/text_language.go [hunk 1]
```diff
@@ -45,35 +43,34 @@ func (p *TextConstraintLanguagePlugin) Validate(dconstraints interface{}) (bool,
...
+				} else if ctrlOp == "(" {
+					parenCount++
+				} else {
+					foundOp = true
+				}
 			}
-
+		}
+		if err == nil && parenCount != 0 {
+			return false, nil, fmt.Errorf(msgPrinter.Sprintf("The constraint expression contains unmatched paretheses."))
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: missing 'n' in paretheses

File: externalpolicy/text_language/text_language_test.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ func Test_Validate_Failed3(t *testing.T) {
 		t.Errorf("Validation should fail but did not, err: %v", err)
 	} else if err == nil {
 		t.Errorf("Validation should fail and return err, but didn't")
-	} else if err.Error() != "Logical operator load is not valid, expecting AND, OR, &&, ||" {
+	} else if err.Error() != "No control operator found:  load && certification == USDA" {
```
Comment (dabooz) [inline externalpolicy/text_language/text_language_test.go]: make check has 3 failures, the first one is here

File: externalpolicy/text_language/text_language.go [hunk 3]
```diff
@@ -85,287 +82,189 @@ func (p *TextConstraintLanguagePlugin) Validate(dconstraints interface{}) (bool,
...
+		nextRune = nextToken.Type
+		if nextRune != def["OpEq"] && nextRune != def["OpComp"] && nextRune != def["OpIn"] {
+			if len(name) > 3 && name[len(name)-2:] == "in" {
+				op = "in"
+				opType = def["in"]
+				name = name[:len(name)-2]
+				val = nextToken.Value
+				valType = nextRune
+			} else {
+				return "", expression, fmt.Errorf("Non-operator token proceeding property name token. \"%s\"", nextToken.Value)
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: Shouldn't this say following property name token?

File: externalpolicy/text_language/text_language.go [hunk 4]
```diff
@@ -85,287 +82,189 @@ func (p *TextConstraintLanguagePlugin) Validate(dconstraints interface{}) (bool,
...
+	// Start of property expression. This case will consume the entire expression.
+	if nextRune == def["Str"] {
+		name := nextToken.Value
+		var opType rune
+		var valType rune
+		op := ""
+		val := ""
+		nextToken, err = lex.Next()
+		if err != nil {
+			return "", expression, fmt.Errorf("Unrecognized token found: %v", err)
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: What causes this error to occur?

File: externalpolicy/text_language/text_language.go [hunk 5]
```diff
@@ -85,287 +82,189 @@ func (p *TextConstraintLanguagePlugin) Validate(dconstraints interface{}) (bool,
...
-	value := pieces[2]
+	  AndOp = whitespace ("AND" | "&&") whitespace .
+	  OrOp = whitespace ("OR" | "||") whitespace .
 
-	// if will failed on case when string values that contain spaces but not quoted (starting from 2nd interation)
-	if !isAllowedComparisonOpType(pieces[1]) {
-		return false, errors.New(msgPrinter.Sprintf("Expression: %v should contain valid comparison operator - wrong operator %v. Allowed operators: %v, %v, %v, %v, %v, %v, %v, %v", expression, pieces[1], lessthan, greaterthan, equalto, doubleequalto, lessthaneq, greaterthaneq, notequalto, inoperator))
-	}
+	  OpEq =  [whitespace]  ( "="["="] | "!=" )   .
+	  OpComp =  [whitespace] ("<"["="] | ">"["="]) [whitespace] .
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: Tweak this to support either >= or =>, same for less than.

File: externalpolicy/text_language/text_language.go (hunk 1 above)
Comment (MaxMcAdam) [reply externalpolicy/text_language/text_language.go]: Fixed

File: externalpolicy/text_language/text_language.go (hunk 3 above)
Comment (MaxMcAdam) [reply externalpolicy/text_language/text_language.go]: Yes. Fixed

File: externalpolicy/text_language/text_language.go (hunk 4 above)
Comment (MaxMcAdam) [reply externalpolicy/text_language/text_language.go]: When the lexer encounters something in the constraint that does not match any of  the specified tokens.

File: externalpolicy/text_language/text_language.go (hunk 5 above)
Comment (MaxMcAdam) [reply externalpolicy/text_language/text_language.go]: Done

File: externalpolicy/text_language/text_language_test.go (hunk 2 above)
Comment (MaxMcAdam) [reply externalpolicy/text_language/text_language_test.go]: Sorry changed error messages after final make check run. Fixed

**Pull request merged**